### ISSUE TITLE
MATLAB catastrophic crashes

### ISSUE THREAD

**Issue by k1o0** [dendritic/signals#17]
Debugging protocols is difficult as MATLAB often suffers a fatal error with no useful information to help solve the issue.  I've noticed that there are a number of TODOs in the signals C code, all pertaining to better error handling.  Hopefully fixing these will help reduce the fatal MATLAB errors.

- [ ] handle allocation fails etc in sqAddNode
- [ ] trap and report errors when notifying event targets in sqApply
- [ ] ensure idx >= 0 && id < net.nNodes && net.nodes[id].inUse in nodePtrs
- [ ] check size and data type of src and dst arrays match in setNodeCurrValue
- [ ] recompile MEX files with debug symbols(?)

**Comment by jaib1** [comment]
fatal crash when trying this assignment to the visStim subscriptable signal handler in an exp def:

```
startTime = t.at(events.expStart.map(true)); 
curExpTime = t - startTime;
circle = vis.patch(t, 'circle');
circle.azimuth = 3 * curExpTime; % this assignment causes the crash
visStim.circle = circle;
```
maybe because trying to render the azimuth too quickly?

**Comment by jaib1** [comment]
This is occurring because the 'azimuth' property for the 'vis.patch' signal is trying to be set to an empty value. This occurs in the above code since 't' does not get posted to until after 'events.expStart' gets posted to. This should be documented and a fix should be implemented.

**Comment by k1o0** [comment]
@jaib1 What you're talking about may be related to cortex-lab/Rigbox#44.  Are you running in a rig with a pre-delay > 0?

**Comment by jaib1** [comment]
yes, I think they may be related, though I was not using a pre-delay. I also believe I've had similar crashes when trying to update a visual stimuli via a signal that does not take a value - I'll try to replicate this. 

Regarding 't' and 'events.expStart', do you know why 'startTime' in this assignment
`startTime = t.at(events.expStart.map(true));`
does not take a value?

**Comment by jaib1** [comment]
> Also note that the main loop begins (and therefore t updates) before expStart updates.

I don't think this is true without a preDelay.

**Comment by jaib1** [comment]
also, assigning an origin signal to a visStim handler element causes a MATLAB fatal crash. e.g:

`os = t.Node.Net.origin('os');`
`os.post(0);`
`rect = vis.patch(t, 'rectangle')`
`rect.azimuth = os;`
`visStim.rect = rect;`

**Comment by k1o0** [comment]
> Regarding 't' and 'events.expStart', do you know why 'startTime' in this assignment
> `startTime = t.at(events.expStart.map(true));`
> does not take a value?

https://github.com/dendritic/signals/blob/c2cf2854491525905ed9f21e0d708b451bfce988/%2Bsig/%2Btransfer/at.m#L17-L28
It's exactly as you say: at the time `events.expStart` updates `t` has no value, so how could it update?  `startTime` would update if you had non-zero pre-delay.



> 
> 
> also, assigning an origin signal to a visStim handler element causes a MATLAB fatal crash. e.g:
> 
> `os = t.Node.Net.origin('os');`
> `os.post(0);`
> `rect = vis.patch(t, 'rectangle')`
> `rect.azimuth = os;`
> `visStim.rect = rect;`

Yes I can reproduce it with the following code:

```
clear all
[t, setElemsFun] = sig.playgroundPTB;
sigbox = t.Node.Net;
rect = vis.patch(t,'rect');
os = sigbox.origin('os');
rect.azimuth = os;
os.post(0.0)
setElemsFun(struct('rect', rect));

rect.show=true;
os.post(1)
```
```
Caught "std::exception" Exception message is:
FatalException
```
Posting to an origin signal before creating dependent signals is pointless: that value won't be propagated.  At first I thought that was the cause of your error.  For me MATLAB only crashes when posting to the origin signal _after_ show is set to true.

**Comment by jaib1** [comment]
> Posting to an origin signal before creating dependent signals is pointless: that value won't be propagated. At first I thought that was the cause of your error

Right, which is why in my code snippet I didn't use a dependent signal based on 'os', I used 'os' directly. Just to be sure, this makes sure this wouldn't be the cause of the error, right? 

> For me MATLAB only crashes when posting to the origin signal _after_ show is set to true.

Hm, are you sure it's only after? That's a bit surprising. Any thoughts on why this would be different for us? What version of MATLAB are you running on your laptop?

**Comment by k1o0** [comment]
Are you using your test version of SignalsExp, if so maybe that's the difference?  I'm using 2018b.  If I run the following I don't have any problems:
```
clear all
[t, setElemsFun] = sig.playgroundPTB;
sigbox = t.Node.Net;
rect = vis.patch(t,'rect');
os = sigbox.origin('os');
rect.azimuth = os;
os.post(0.0)
setElemsFun(struct('rect', rect));
% Start timer %
os.post(0.0)
os.post(1)
os.post('h')
os.post(3)
```

**Comment by jaib1** [comment]
But in that code snippet, you are not setting:
`rect.show = true`

There are a few issues that make it impossible for me to run 'sig.playgroundPTB' in it's current form (incompatibility with my version of GUI layout toolbox, incompatibility with how the screens are defined, etc...), but if you run this exact expdef

```
function visTest(t, events, params, visStim, inputs, outputs, audio)
%VISTEST For testing visual stimuli   

  os = t.Node.Net.origin('os');
  os.post(1);
  rect = vis.patch(t, 'rectangle');
  rect.azimuth = os;
  rect.show = true;
  visStim.rect = rect;
  events.endTrial = events.newTrial.delay(5);

end
```
and it doesn't crash, then something very strange indeed is happening. For me, the crash occurs regardless of where or whether or not I post to 'os' in the expdef, simply as long as 'os' is defined as an origin signal and fed into one of the elements of 'rect', and 'rect.show' is set to 'true'.

**Comment by k1o0** [comment]
RE sig.playgroundPTB: Could you create an issue and address these issues as a matter of urgency (create a pull request; don't hotfix)?  We need playgroundPTB to work smoothly on most computers as it's used in the Ringach example and should be used in other GUI-based examples.

**Comment by k1o0** [comment]
> 
> 
> But in that code snippet, you are not setting:
> `rect.show = true`

Yes, as I said earlier MATLAB only crashed when posting to the signal after the stimulus is shown.  If I run the above code I **don't** get a crash.  The only difference if that show is set to false.

**Comment by k1o0** [comment]
When I run the code posted [here](https://github.com/dendritic/signals/issues/17#issuecomment-481220939) MATLAB doesn't crash.  If I then type `rect.show = true;`, the next time I update the origin signal MATLAB crashes.

**Comment by jaib1** [comment]
and does it crash if you run this:
```
clear all
[t, setElemsFun] = sig.playgroundPTB;
sigbox = t.Node.Net;
rect = vis.patch(t,'rect');
os = sigbox.origin('os');
os.post(0.0)
rect.azimuth = os;
rect.show = true;
setElemsFun(struct('rect', rect));
```
and this:
```
clear all
[t, setElemsFun] = sig.playgroundPTB;
sigbox = t.Node.Net;
rect = vis.patch(t,'rect');
os = sigbox.origin('os');
rect.azimuth = os;
rect.show = true;
setElemsFun(struct('rect', rect));
```
?

**Comment by k1o0** [comment]
Yes

**Comment by jaib1** [comment]
also occasionally get crashes when running `clear all` in the command window where a network and signals have been defined.

**Comment by jaib1** [comment]
'onValue' causes MATLAB crashes occasionally:

e.g. 
```
net = sig.Net;
os = net.origin('os');
l = feval(@onValue, os, @(x) disp(x));
os1.post(1);
```
(I realize this is not typically the syntax used, but this syntax is what caused the crash)
